packages/transformers/js/src/JSTransformer.js

```diff
@@ -706,51 +706,73 @@ export default (new Transformer({
-      );
-      for (let dep of deps.values()) {
-        dep.symbols.ensure();
+      // deps is a Map of specifiers/placeholders to an array of dependencies
+      // since multiple dependencies can be attributed to a single specifier
+      let deps = new Map();
+      for (let dep of asset.getDependencies()) {
+        if (!deps.has(dep.meta.placeholder ?? dep.specifier)) {
+          deps.set(dep.meta.placeholder ?? dep.specifier, []);
+        }
+        deps.get(dep.meta.placeholder ?? dep.specifier)?.push(dep);
+      }
+      for (let depArr of deps.values()) {
+        for (let dep of depArr) {
+          dep.symbols.ensure();
+        }
       }
 
       for (let {
         source,
         local,
         imported,
         loc,
+        kind,
       } of hoist_result.imported_symbols) {
-        let dep = deps.get(source);
-        if (!dep) continue;
-        dep.symbols.set(imported, local, convertLoc(loc));
+        for (let dep of nullthrows(deps.get(source))) {
+          // dep.meta.kind is a DependencyKind while kind is an ImportKind.
+          // We only set the symbols for the dep if both dep.meta.kind is an
+          // ImportKind and if dep.meta.kind and kind are the same, or if
+          // dep.meta.kind is not an ImportKind
+          if (
+            (dep.meta.kind === 'Require' ||
+              dep.meta.kind === 'Import' ||
+              dep.meta.kind === 'DynamicImport') &&
+            dep.meta.kind !== kind
+          ) {
+            continue;
+          }
+          dep.symbols.set(imported, local, convertLoc(loc));
+        }
       }
-
       for (let {source, local, imported, loc} of hoist_result.re_exports) {
-        let dep = deps.get(source);
-        if (!dep) continue;
-
-        if (local === '*' && imported === '*') {
-          dep.symbols.set('*', '*', convertLoc(loc), true);
-        } else {
-          let reExportName =
-            dep.symbols.get(imported)?.local ??
-            `$${asset.id}$re_export$${local}`;
-          asset.symbols.set(local, reExportName);
-          dep.symbols.set(imported, reExportName, convertLoc(loc), true);
+        for (let dep of nullthrows(deps.get(source))) {
+          if (dep.specifierType !== 'esm') {
+            continue;
+          }
+          if (local === '*' && imported === '*') {
+            dep.symbols.set('*', '*', convertLoc(loc), true);
+          } else {
+            let reExportName =
+              dep.symbols.get(imported)?.local ??
+              `$${asset.id}$re_export$${local}`;
+            asset.symbols.set(local, reExportName);
+            dep.symbols.set(imported, reExportName, convertLoc(loc), true);
+          }
         }
       }
 
       for (let specifier of hoist_result.wrapped_requires) {
-        let dep = deps.get(specifier);
-        if (!dep) continue;
-        dep.meta.shouldWrap = true;
+        for (let dep of nullthrows(deps.get(specifier))) {
+          dep.meta.shouldWrap = true;
+        }
       }
```

Comment:
`dep.meta.shouldWrap` will get set for every dependency with that specifier. Even if it should have been requires only.